src/migrate.js

```diff
@@ -36,161 +32,102 @@ function migrate(dbConfig = {}, migrationsDirectory, config = {}) { // eslint-di
 
   log("Attempting database migration")
 
+  let executedMigrations = null
+
   try {
-    return bluebird.resolve()
-      .then(() => client.connect())
-      .then(() => log("Connected to database"))
-      .then(() => loadMigrationFiles(migrationsDirectory, log))
-      .then(filterMigrations(client))
-      .each(runMigration(client))
-      .then(finalise(client, log))
-      .catch((err) => {
-        log(`Migration failed. Reason: ${err.message}`)
-        try {
-          return client.end()
-            .then(() => {
-              throw err
-            })
-            .catch(() => {
-              throw err
-            })
-        } catch (e) {
-          return Promise.reject(err)
-        }
-      })
-  } catch (e) {
-    return Promise.reject(e)
-  }
-}
+    await client.connect()
+    log("Connected to database")
 
-function finalise(client, log) {
-  return (completedMigrations) => {
-    if (completedMigrations.length === 0) {
-      log("No migrations applied")
-    } else {
-      const names = completedMigrations.map((m) => m.name)
-      log(`Successfully applied migrations: ${names}`)
-    }
-
-    return client.end()
-      .then(() => completedMigrations)
-  }
-}
+    const migrations = await filesLoader.load(migrationsDirectory, log)
 
-// Work out which migrations to apply
-function filterMigrations(client) {
-  return (migrations) => {
-    // Arrange in ID order
-    const orderedMigrations = migrations.sort((a, b) => a.id - b.id)
-
-    // Assert their IDs are consecutive integers
-    migrations.forEach((mig, i) => {
-      if (mig.id !== i) {
-        throw new Error("Found a non-consecutive migration ID")
-      }
-    })
-
-    return doesTableExist(client, "migrations")
-      .then((exists) => {
-        if (!exists) {
-          // Migrations table hasn't been created,
-          // so the database is new and we need to run all migrations
-          return orderedMigrations
-        }
-
-        return client.query("SELECT * FROM migrations")
-          .then(filterUnappliedMigrations(orderedMigrations))
-      })
+    const appliedMigrations = await fetchAppliedMigrationFromDB(migrationTableName, client, log)
+
+    validateMigrations(migrations, appliedMigrations)
+
+    const filteredMigrations = filterMigrations(migrations, appliedMigrations)
+
+    const completedMigrations = await Promise.all(filteredMigrations.map(runMigration(migrationTableName, client, log)))
+
+    executedMigrations = finalize(completedMigrations, log)
+  } catch (err) {
+    log(`Migration failed. Reason: ${err.message}`)
+    throw new Error(dedent`
+      ${err.message}
+      ${err.stack}
```

Comment:
What does this error message look like? I'm not sure I like the idea of putting the stack in the message. Maybe something like this would be better:

```js
const error = new Error(`Migration failed. Reason: ${err.message}`)
error.cause = err
throw error
```